pkg/tests/tasks/security/certificate/external_ca_test.go

```diff
@@ -16,101 +16,116 @@ package certificate
-			defer util.CloseResponseBody(resp)
-		} else {
-			log.Log.Info("Verify the new certificates")
-
-			// Generate the cert files
-			util.ShellMuteOutput(`oc -n bookinfo exec %s -c istio-proxy -- openssl s_client -showcerts -connect details:9080 > %s/bookinfo-proxy-cert.txt`, productPod, tmpDir)
-			_, err = util.ShellMuteOutput(`sed -n '/-----BEGIN CERTIFICATE-----/{:start /-----END CERTIFICATE-----/!{N;b start};/.*/p}' %s/bookinfo-proxy-cert.txt > %s/certs.pem`, tmpDir, tmpDir)
-			util.Inspect(err, "Failed to parse 'openssl s_client' output", "", t)
-			_, err = util.ShellMuteOutput(`awk 'BEGIN {counter=0;} /BEGIN CERT/{counter++} { print > "%s/proxy-cert-" counter ".pem"}' < %s/certs.pem`, tmpDir, tmpDir)
-			util.Inspect(err, "Failed to split certs into separate files", "", t)
-
-			// Compare them with the original certs
-			log.Log.Info("Verifying the root certificate")
-			_, err = util.ShellMuteOutput(`openssl x509 -in %s -text -noout > %s/root-cert.crt.txt`, sampleCARoot, tmpDir)
-			util.Inspect(err, "Failed to print cert", "", t)
-			_, err = util.ShellMuteOutput(`openssl x509 -in %s/proxy-cert-3.pem -text -noout > %s/pod-root-cert.crt.txt`, tmpDir, tmpDir)
-			util.Inspect(err, "Failed to print cert", "", t)
-			if err := util.CompareFiles(fmt.Sprintf("%s/root-cert.crt.txt", tmpDir), fmt.Sprintf("%s/pod-root-cert.crt.txt", tmpDir)); err != nil {
-				t.Errorf("Root certs do not match: %v", err)
-			}
+			t.LogStep("Checking response from productpage.")
+			retry.UntilSuccess(t, func(t test.TestHelper) {
+				curl.Request(t, app.BookinfoProductPageURL(t, meshNamespace), curl.NilRequestOption{}, assert.ResponseStatus(200))
+			})
 
-			log.Log.Info("Verifying the CA certificate")
-			_, err = util.ShellMuteOutput(`openssl x509 -in %s -text -noout > %s/ca-cert.crt.txt`, sampleCACert, tmpDir)
-			util.Inspect(err, "Failed to print cert", "", t)
-			_, err = util.ShellMuteOutput(`openssl x509 -in %s/proxy-cert-2.pem -text -noout > %s/pod-cert-chain-ca.crt.txt`, tmpDir, tmpDir)
-			util.Inspect(err, "Failed to print cert", "", t)
-			if err := util.CompareFiles(fmt.Sprintf("%s/ca-cert.crt.txt", tmpDir), fmt.Sprintf("%s/pod-cert-chain-ca.crt.txt", tmpDir)); err != nil {
-				t.Errorf("CA certs do not match: %v", err)
-			}
+		} else {
 
-			log.Log.Info("Verifying the certificate chain")
-			output, err := util.ShellMuteOutput(`/bin/bash -c "openssl verify -CAfile <(cat %s %s) %s/proxy-cert-1.pem"`, sampleCACert, sampleCARoot, tmpDir)
-			util.Inspect(err, "Failed to verify the certificate chain", "", t)
-			expected := []byte(fmt.Sprintf("%s/proxy-cert-1.pem: OK", tmpDir))
-			if err := util.Compare([]byte(output), expected); err != nil {
-				t.Errorf("unexpected output while verifying cert chain: %v", err)
-			}
+			t.LogStep("Retrieving certificates.")
+			tmpDir = shell.CreateTempDir(t, "cacerts")
+			oc.Exec(t, pod.MatchingSelector("app=productpage", "bookinfo"), "istio-proxy",
```

Comment:
We really should re-think `oc.Exec`. I need to redirect the output of kubectl locally, but in order to do that I am passing in the output redirection as part of the `command`. Anything in `command` _should_ be executing on the remote pod, and things like checking the kubectl return value (the reason I needed the `|| true`) and local output redirection should be handled via some other mechanism.

Reply:
I agree. I've just pushed a change to this PR to make `oc.Exec` return the output. You can then write this output to a file using `util.WriteTextFile()` (but please create a helper that calls this file and fails the test if an error occurs instead of adding error handling code to this test).